File: rate_limit/automation/lib/python/community/rate_limit.py
"""
Copyright June 25, 2020 Richard Koshak

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
from community.time_utils import to_datetime
from org.joda.time import DateTime
#from datetime import datetime, timedelta

logger = logging.getLogger("{}.Rate Limit".format(LOG_PREFIX))

class RateLimit(object):
    """Keeps a timestamp for when a new call to run is allowed to execute, ignoring
    any calls that take place before that time.
    """

    @log_traceback
    def __init__(self):
        """ Initializes the timestamp to now. """

        self.until = DateTime().now().minusSeconds(1)

    @log_traceback
    def run(self, func, when):
        """If it has been long enough since the last time that run was called,
        execute the passed in func. Otherwise ignore the call.
        Arguments:
            - func: The lambda or function to call if allowed.
            - when: When the rate limit will expire. Can be a DateTime type
            object, a number which is treated as seconds, or a duration string
            (e.g. 5m 2.5s), or an ISO 8601 formatted date string. See time_utils
            for details
        """

        now = DateTime().now()
        if now.isAfter(self.until):
            self.until = to_datetime(when)
            func()

File: rate_limit/javascript/automation/lib/javascript/community/rateLimit.js
/**
 * A class that will limit how often an event can occur. One calls run and pass
 * a time_utils when to indicate how long before the call to run will run again.
 * If run is called before that amount of time then the call is ignored.
 */
 var RateLimit = function() {
    'use strict';
  
    var OPENHAB_CONF = java.lang.System.getenv("OPENHAB_CONF");
    this.ZonedDateTime = Java.type("java.time.ZonedDateTime");
    this.log = Java.type("org.slf4j.LoggerFactory").getLogger("org.openhab.model.script.Rules.RateLimit");
    this.log.debug("Building the RateLimit object.");
    load(OPENHAB_CONF+'/automation/lib/javascript/community/timeUtils.js');
    this.until = this.ZonedDateTime.now().minusSeconds(1);
    this.log.debug("RateLimit is ready to operate");
  }
  
  /**
   * Function called to attempt to run the passed in function. If enough time has
   * passed since the last time run was called func is called. If not the call is
   * ignored.
   * 
   * @param {function} func called if it's been long enough since the last call to run
   * @param {*} when any of the durations supported by time_utils.
   */
  RateLimit.prototype.run = function(func, when){
    var now = this.ZonedDateTime.now();
    if(now.isAfter(this.until)) {
      this.until = toDateTime(when);
      func();
    }
  }